SEP-0012.md

```diff
@@ -0,0 +1,495 @@
+In addition to this in-between the release of ndcube 1.0 and the development of 2.0, various long running projects such as the acceptance and implementation of [APE 14](https://github.com/astropy/astropy-APEs/blob/master/APE14.rst) were completed.
+This means that once many of these brand new features had been implemented in Astropy and gWCS it provided an excellent opportunity to reassess the state of NDCube and to rebase it onto this new WCS functionality.
+
+The main motivation for updating ndcube to use the APE 14 interface was to provide support for gWCS which is being used by the DKIST user tools.
+A lot of the other changes presented in this SEP are a consequence of this decision.
+
+
+## Summary of Changes from ndcube 1
+
+The fundamental change since the original version of ndcube is that the WCS type has been changed to be the High Level API as described by astropy's APE 14.
+This opens a lot of possibilities to the future uses of `NDCube` as the APE 14 API allows the coordinate system to be described by almost any functional transform.
+This could be the most common FITS WCS standard, or gWCS or custom implementations for specific facilites with different requirements.
+
+Having made this change, which was a large breaking change from the original API, we reviewed all the existing functionality and API as it was adapted to use the high level WCS interface and the result is presented below.
+
+# Detailed Description
+
+The API presented here is limited to the two core classes `NDCubeABC` and `ExtraCoordsABC`.
+This is primarily an attempt to limit the discussion in this SEP to the core API and not implementation decisions taken in the development of ndcube.
+If people are interested in the implementation of the proposed API large amounts of it are implemented on the master branch of the [ndcube repository](https://github.com/sunpy/ndcube) or open pull requests.
+
+The `NDCube` class inherits the astropy [`NDDataBase`](https://github.com/astropy/astropy/blob/master/astropy/nddata/nddata_base.py#L11) class, including the requirement that the `.wcs` property be a `BaseHighLevelWCS` instance.
+In the following code snippets of the API use the Python type hint syntax as appropriate to highlight accepted types and return types, these are not enforced at runtime.
+
+
+## NDCube
+
+Below is a class definition for the Abstract Base Class (ABC) for `NDCube`.
+Its purpose is to specify the API as completely as possible, including input and return types. 
+The behaviour of all the methods should be described by their included docstrings, a version of this (likely out of sync) has been [PRed to ndcube](https://github.com/sunpy/ndcube/pull/304).
+
+```python
+
+class NDCubeABC(astropy.nddata.NDDataBase):
+
+    @abc.abstractmethod
+    def __init__(self,
+                 data: "numpy.typing.ArrayLike",
+                 uncertainty: Any = None,
```

Comment:
Why can this be `Any` and not just a numpy array or `None`?

Reply:
Also how does one take into account that uncertainty could have different units to the data (e.g. data could be in `m`, and uncertainty in `cm`).

Reply:
This should ideally be the nddata uncertainty type, but nddata doesn't actually *enforce* that type, so `Any` 
is technically correct. `NDCube` doesn't currently do any handling of this at all, it's all done by the base class(es).

Uncertainty handling in astropy is functional but limited, hopefully with this API we can transparently benefit from any improvements made upstream.